alf/algorithms/muzero_algorithm.py

```diff
@@ -295,70 +295,182 @@ def _hook(grad, name):
-                    r = torch.randperm(
-                        batch_size) < batch_size * self._reanalyze_ratio
+                    r = torch.randperm(B) < B * self._reanalyze_ratio
+                    # [B', T + R, ...], B' = B * reanalyze_ratio
                     r_candidate_actions, r_candidate_action_policy, r_values = self._reanalyze(
-                        replay_buffer, env_ids[r], positions[r],
-                        mcts_state_field)
+                        replay_buffer, start_env_ids[r], start_positions[r],
+                        mcts_state_field, T + R - 1)
+                    # [B', T, R + 1, ...]
+                    r_candidate_actions, r_candidate_action_policy, r_values = alf.nest.map_structure(
+                        _unfold_dim1_adapting_episode_ends,
+                        (r_candidate_actions, r_candidate_action_policy,
+                         r_values))
                 else:
+                    # [B, T + R, ...]
                     candidate_actions, candidate_action_policy, values = self._reanalyze(
-                        replay_buffer, env_ids, positions, mcts_state_field)
-
-            # [B]
-            steps_to_episode_end = replay_buffer.steps_to_episode_end(
-                positions, env_ids)
-            # [B]
-            episode_end_positions = positions + steps_to_episode_end
-
-            # [B, unroll_steps+1]
-            positions = positions.unsqueeze(-1) + torch.arange(
-                self._num_unroll_steps + 1)
-            # [B, 1]
-            env_ids = batch_info.env_ids.unsqueeze(-1)
-            # [B, 1]
-            episode_end_positions = episode_end_positions.unsqueeze(-1)
+                        replay_buffer, start_env_ids, start_positions,
+                        mcts_state_field, T + R - 1)
 
-            beyond_episode_end = positions > episode_end_positions
-            positions = torch.min(positions, episode_end_positions)
+                    # [B, T, R + 1, ...]
+                    candidate_actions, candidate_action_policy, values = alf.nest.map_structure(
+                        _unfold_dim1_adapting_episode_ends,
+                        (candidate_actions, candidate_action_policy, values))
 
             if self._reanalyze_ratio < 1:
                 if self._td_steps >= 0:
+                    # [B, T + R]
                     values = self._calc_bootstrap_return(
-                        replay_buffer, env_ids, positions, value_field)
+                        replay_buffer, folded_env_ids, folded_positions,
+                        value_field)
                 else:
+                    # [B, T + R]
                     values = self._calc_monte_carlo_return(
-                        replay_buffer, env_ids, positions, value_field)
+                        replay_buffer, folded_env_ids, folded_positions,
+                        value_field)
+
+                # [B, T, R + 1]
+                values = _unfold_dim1(values)
```

Comment:
This `values` should also be applied with `_unfold_dim1_adapting_episode_ends`

Reply:
Fixed.